### DISCUSSION TITLE
Dependency Injection for startup in test client

### DISCUSSION THREAD

**Discussion by scottherlihy** [fastapi/fastapi#6443] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# currently:

@app.on_event("startup")
async def startup():
    load_defaults()
    await DB.start_db(settings=settings)
    return

# what I'd like
@app.on_event("startup")
async def startup(defaults_retriever=None):
    if not defaults_retriever:
        defaults_retriever = RealRetriever()
    load_defaults(defaults_retriever)
    await DB.start_db(settings=settings)
    return



# and then in tests I wouldnt need to mock the remote because I can do something like 
client = TestClient(app, startup_args=[MockDefaultsRetriever])
```

### Description

Is it possible to pass an argument to the startup function of the TestClient?

https://fastapi.tiangolo.com/advanced/testing-events/

The docs suggest using TestClient to ensure the startup function gets invoked, but what if the startup function has an argument that is a class that hits a remote and Id like to mock it. Since patching module can be finicky  dependency injection is a clean way to accomplish the goal of mocking an object and if something like I suggested in the example code was implemented it would make this very easy.

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

Python 3.9.7

**Comment by DavidBord** [comment]
I think that if that would have been made possible then all the startup & shutdown events would have gotten these args. I wouldn't have used this feature if it was available.
Would you consider using `partial` from `functools` for your use case?

**Comment by a-n-d-i** [comment]
Hi, maybe just mock/patch RealRetriever? Or call the decorator directly? I mean you're already passing an Argument to Testclient, namely app. The startup stuff is just buried in there. So doing such an Argument list would be like passing the same stuff twice to avoid changing the insides of the app parameter.

**Comment by adriangb** [comment]
I think this duplicates the feature request in https://github.com/tiangolo/fastapi/issues/617

**Comment by scottherlihy** [comment]
I agree, @adriangb. Closing.
